review-problems.ml

```diff
@@ -917,3 +917,203 @@ module Lazy = struct
+
+  - Implement a function `average_grade` with the following signature:
+
+      val average_grade : student -> float
+
+      This function should take a single student and calculate their average grade across all courses. 
+      Use this function to find the student with the highest average grade in the list of students.
+
+  - Implement a function `top_students` with the following signature:
+
+      val top_students : student list -> name list
+
+      This function should take a list of students and return a list of names of students who have achieved a 
+      grade of 95 or higher in at least one course.
+
+*)
+
+(*
+QUESTION#5 :-
+
+Recursive Chessboard Exploration
+
+Consider an `n` x `n` chessboard represented as a grid of squares. Each square can be either empty or blocked. 
+You start at the top-left square `(0, 0)` and want to reach the bottom-right square `(n-1, n-1)`. 
+However, there are some restrictions:
+
+- You can only move right or down.
+- You cannot move through blocked squares.
+- You cannot visit the same square twice.
+
+(1). Implement a recursive function `explore_chessboard` with the following signature:
+
+  val explore_chessboard : int -> bool array array -> bool
+
+      This function should take an integer `n` (the size of the chessboard), a 2D boolean array representing the 
+      blocked squares (where `true` indicates a blocked square), and return `true` if there is a path from the top-left 
+      corner `(0, 0)` to the bottom-right corner `(n-1, n-1)` that adheres to the rules mentioned above. Otherwise, return `false`.
+
+(2). Create a higher-order function `process_path` with the following signature:
+
+  val process_path : (int * int -> unit) -> int -> bool array array -> unit
+
+      This function should take a function `process` that accepts coordinates `(x, y)` and a 2D boolean array representing the chessboard. 
+      It should apply the `process` function to each square in the path from `(0, 0)` to `(n-1, n-1)` while adhering to the rules, 
+      starting from the top-left corner and ending at the bottom-right corner.
+
+(3). Implement a function `print_chessboard` with the following signature:
+
+  val print_chessboard : bool array array -> unit
+
+      This function should take a 2D boolean array representing the chessboard and print it to the console, using `X` to represent 
+      blocked squares and `.` to represent empty squares.
+
+(4). Write a program that demonstrates the usage of these functions. Create a chessboard, print it, explore it using the `explore_chessboard` function
+and print the path if a valid path exists.
+
+
+*)
+
+(*
+
+QUESTION#6 :-
+
+You are presented with an encoded message consisting of characters and instructions for decryption. 
+
+Your task is to implement a decoding algorithm following the provided instructions.
+
+(1). Implement a recursive function `decode_message` with the following signature:
```

Comment:
This is a good problem but it needs more scaffolding. These questions are intended to be suitable as practice for a midterm, so they should not require the students to look up lots of library functions. Rather, minimal library functions should be required to implement the solution, and the required ones should be specified in the question.

Most of the implementations you wrote as solutions are converting to/from lists/sequences to then work with lists, so the problem can be adjusted to just operate on lists of characters and hence avoid all these conversions.